src/inputs/ToggleButton.tsx

```diff
@@ -0,0 +1,109 @@
+import { ReactNode, useRef, useState } from "react";
+import { useHover, useSwitch, useFocusRing, VisuallyHidden, useButton } from "react-aria";
+import { isPromise, toToggleState, useTestIds } from "src/utils";
+import { Css } from "src/Css";
+import { Icon, IconKey, maybeTooltip, resolveTooltip } from "..";
+
+export interface ToggleButtonProps {
+  /** Input label */
+  label: string;
+  selected?: boolean;
+  onClick: ((selected: boolean) => void) | ((active: boolean) => Promise<void>);
+  autoFocus?: boolean;
+  icon?: IconKey;
+  /** Whether the field is disabled. If a ReactNode, it's treated as a "disabled reason" that's shown in a tooltip. */
+  disabled?: boolean | ReactNode;
+  tooltip?: ReactNode;
+  /** Handler when the interactive element state changes. */
+}
+
+export function ToggleButton(props: ToggleButtonProps) {
+  const {
+    selected: isSelected = false,
+    disabled = false,
+    label,
+    onClick,
+    icon,
+    ...otherProps
+  } = props;
+  const isDisabled = !!disabled;
+  const state = toToggleState(isSelected, onClick);
+  const [asyncInProgress, setAsyncInProgress] = useState(false);
+
+  const ref = useRef(null);
+  const tid = useTestIds(props, label);
+
+  const ariaProps = { isSelected, isDisabled: isDisabled || asyncInProgress, ...otherProps };
+  const { inputProps } = useSwitch({ ...ariaProps, "aria-label": label }, state, ref);
+  const { buttonProps, isPressed } = useButton(
+    {
+      ...ariaProps,
+      onPress: () => {
+        const result = onClick(!isSelected);
+        if (isPromise(result)) {
+          setAsyncInProgress(true);
+          result.finally(() => setAsyncInProgress(false));
+        }
+        return result;
+      },
+      elementType: "button",
+    },
+    ref,
+  );
+  const { isFocusVisible: isKeyboardFocus, focusProps } = useFocusRing(otherProps);
+  const { hoverProps, isHovered } = useHover(ariaProps);
+
+  const tooltip = resolveTooltip(disabled);
+
+  const buttonAttrs = {
+    ref: ref as any,
+    ...buttonProps,
+    ...focusProps,
+    ...hoverProps,
+    css: {
+      ...Css.buttonBase.br8.tt("inherit").$,
+      ...Css.bgTransparent.gray500.hPx(32).pxPx(12).$,
+      ...(isHovered && toggleHoverStyles),
+      ...(isPressed ? togglePressStyles : {}),
+      ...(isDisabled && Css.gray300.$),
+      ...(isSelected && Css.lightBlue700.$),
+      ...(isSelected && isHovered && toggleSelectedHoverStyles),
+      ...(isKeyboardFocus && toggleFocusStyles),
+    },
+    ...tid,
+  };
+
+  return maybeTooltip({
+    title: tooltip,
+    placement: "top",
+    children: (
+      <label
+        {...hoverProps}
+        css={{
+          ...Css.relative.cursorPointer.df.w("max-content").smEm.selectNone.$,
+          ...(isDisabled && Css.cursorNotAllowed.gray400.$),
+        }}
+        aria-label={label}
+      >
+        <button
+           aria-hidden="true"
+          {...buttonAttrs}
+        >
```

Comment:
**thought** I am thinking we don't really want a button element here. As it is right now, we have nested interactions. When a user clicks on the `<button />` element, then we fire the `onPress`, then that click bubbles up to the `<label/>` which toggle the state of the `<input />`. 
I would suggest we get rid of the button and `useButton`. The label should be sufficient for our click interaction.

Reply:
Yeah, I tried this at first, but the problem is that `useSwitch` doesn't support a `onPress` so we won't have the `isPressed` state, which is minor thing but I was just following the figma design

Reply:
Also, the async behavior doesn't seem to work properly when applying the same logic on the `onChange`

Reply:
Hmm.. I think we can still get around this. Let me pull this down and take a look as well.

Reply:
@bdow pushed!